### PR TITLE
Add json-ld to work/books pages

### PR DESCRIPTION
**Pull request by Sabreen-Parveen**
Closes #
This PR is in reference to #4182 

It adds JSON-LD feature in work/books pages which helps in better search.

### Technical

I have not included the `potentialAction` part.
### Screenshot
For the book **The interior castle**, it shows following result:
a
b

### Stakeholders

@cdrini

### CODE REVIEW

File: openlibrary/templates/type/edition/view.html [hunk 1]
```diff
@@ -497,3 +502,57 @@ <h2 class="home-h2"><a href="$page.url()/lists">$_('Lists containing this Book')
...
+    $return identifier
+
+$def get_url(label):
+    $ url = list(values[0].url for name, values in edition.get_identifiers().multi_items() if values[0].label == label)
+    $ url = url[0] if url else None
+    $return url
+    
+
+<script type="application/ld+json">
+{
```
Comment (cdrini) [inline openlibrary/templates/type/edition/view.html:518]: Constructing the JSON here could be a little error-prone ; we should move this logic to a new function inside the work and edition model (maybe `as_json_ld(work, edition=None)`). Then we can build the json as a python `dict`, and convert it to a JSON string using `json.dumps`. Let me know if you have any questions!

File: openlibrary/templates/type/edition/view.html [hunk 2]
```diff
@@ -497,3 +502,57 @@ <h2 class="home-h2"><a href="$page.url()/lists">$_('Lists containing this Book')
...
+  "@context": "http://schema.org/",
+  "@type": "Book",
+  "@id": "$base_url/$work.key",
+  "url": "$base_url/$work.key/$book_title",
```
Comment (cdrini) [inline openlibrary/templates/type/edition/view.html:522]: Too many slashes here (work.key already has a slash)

```suggestion
  "@id": "$base_url$work.key",
  "url": "$base_url$work.key/$book_title",
```

File: openlibrary/templates/type/edition/view.html [hunk 3]
```diff
@@ -497,3 +502,57 @@ <h2 class="home-h2"><a href="$page.url()/lists">$_('Lists containing this Book')
...
+      "@type": "AggregateRating",
+      "ratingValue": "$rating_stats.get('avg_rating',0)",
+      "bestRating": "5",
+      "ratingCount": "$rating_stats.get('num_ratings', 0)"
+  },
+
+  "workExample": [
+      {
+          "@type": "Book",
+          "@id": "$base_url/$edition.key",
```
Comment (cdrini) [inline openlibrary/templates/type/edition/view.html:535]:
```suggestion
          "@id": "$base_url$edition.key",
```

File: openlibrary/templates/type/edition/view.html [hunk 4]
```diff
@@ -497,3 +502,57 @@ <h2 class="home-h2"><a href="$page.url()/lists">$_('Lists containing this Book')
...
+          "isbn": "$:get_identifier("ISBN 10")",
```
Comment (cdrini) [inline openlibrary/templates/type/edition/view.html:536]: We want to use isbn13 here.

File: openlibrary/templates/type/edition/view.html [hunk 5]
```diff
@@ -38,6 +38,11 @@
 $ edition_count = work.edition_count if work and work.edition_count else 1
 $ meta_cover_url = item_image((edition or work).get_cover_url("L"), default="https://openlibrary.org/images/icons/avatar_book-sm.png")
 $ _x = ctx.setdefault('bodyid', 'book')
+$ base_url = "https://openlibrary.org"
```
Comment (cdrini) [inline openlibrary/templates/type/edition/view.html:41]: I _think_ this should work with `ctx.homedomain` ? But you might want to test it first.

**Review by cdrini — CHANGES_REQUESTED** [review]
This looks like a great start! Main feedback is let's try moving this to the python ; that will make it easier for us to test and to handle some edge cases (like books without Goodreads IDs, e.g.).

We likely want to add a function to the `Work` model and the `Edition` model in https://github.com/internetarchive/openlibrary/blob/85391b39c015d3b5f8b5032c3e101132a2737539/openlibrary/plugins/upstream/models.py#L46 .

File: openlibrary/templates/type/edition/view.html (hunk 1 above)
Comment (Sabreen-Parveen) [reply openlibrary/templates/type/edition/view.html:518]: Where should I call this function? In `type/edition/view.html` or somewhere else?

Comment (cdrini) [reply openlibrary/templates/type/edition/view.html:518]: Yep! Exactly. You should be able to do `work.as_json_ld()` or `edition.as_json_ld()` or `page.as_json_ld()`.

**Comment by Sabreen-Parveen** [comment]
@cdrini for language field, 2 letter codes are recommended, but there are only 3 letter codes in 
`openlibrary/plugins/openlibrary/pages/languages.page`.

image
Can I add 2 letter codes in this file?

**Comment by cdrini** [comment]
For now let's use the three letter codes; converting between the two and three letter codes is actually a lot of work :D We can do that in a future PR.

File: openlibrary/templates/type/edition/view.html [hunk 6]
```diff
@@ -499,3 +504,63 @@ <h2 class="home-h2"><a href="$page.url()/lists">$_('Lists containing this Book')
...
+      }
+  ]
+}
+</script>
+<script type="application/ld+json">
+$if page.key.startswith('/works'):
+    $:work.as_json_ld()
+$elif page.key.startswith('/books'):
+    $:edition.as_json_ld()
+</script>
```
Comment (cdrini) [inline openlibrary/templates/type/edition/view.html:566]:
```suggestion
<script type="application/ld+json">
$if work:
    $:work.as_json_ld(edition)
$else:
    $:edition.as_json_ld()
</script>
```

File: openlibrary/plugins/upstream/models.py [hunk 7]
```diff
@@ -489,6 +490,48 @@ def is_fake_record(self):
...
+                         self.get_cover_url("L")
+                    ],
+                    "bookFormat": "http://schema.org/"+ self.physical_format.replace("[", "").replace("]", "") if self.physical_format else None,
+                    "inLanguage": (", ".join(lang.code for lang in self.languages)) or None,
+                    "datePublished": self.publish_date or None,
+                    "sameAs":[
+                        get_url("OCLC/WorldCat"),
+                        get_url("Library Thing"),
+                        get_url("Goodreads")
+                    ],
```
Comment (cdrini) [inline openlibrary/plugins/upstream/models.py:523]: Avoid missing urls

```suggestion
                    "sameAs": [for url in (
                        get_url("OCLC/WorldCat"),
                        get_url("Library Thing"),
                        get_url("Goodreads")
                    ) if url],
```

File: openlibrary/plugins/upstream/models.py [hunk 8]
```diff
@@ -489,6 +490,48 @@ def is_fake_record(self):
...
+                    "identifier":[
+                        {
+                        "@type": "PropertyValue",
+                        "propertyID": "OCLC_NUMBER",
+                        "value": get_identifier("OCLC/WorldCat")
+                        }
+                    ]
```
Comment (cdrini) [inline openlibrary/plugins/upstream/models.py:530]: Also place this in an if block so we don't include it if the OCLC is not specified.

**Review by cdrini — CHANGES_REQUESTED** [review]
Looking good! Almost there :)

Main points:

- Let's make those json_ld function return `dict`s, and the jsonify them from the template
- A key note is that if a book has a work and a edition, we want to call _both_ the `as_json_ld` functions

File: openlibrary/plugins/upstream/models.py (hunk 7 above)
Comment (Sabreen-Parveen) [reply openlibrary/plugins/upstream/models.py]: The `get_url` function returns None for missing urls.

**Comment by hornc** [comment]
@cdrini, thanks for linking to this. I have been looking at OL's RDF and trying to figure out what is needed to make it up to scratch to function in the [LOD Cloud](https://lod-cloud.net/), to link with Wikidata and VIAF etc.

A few things from my investigation so far around URIs vs. URLs:

I believe the correct form of Open Library concept URIs is the `http` version (without a trailing slash) i.e.  `http://openlibrary.org/works/OL6037022W`

We should use some form of content negotiation to get to a relevant data URL from there, something like what was proposed in #130, but updated since some the current redirect behaviour has changed since then (there are still issues, but new ones) 

And similarly the URIs for World Cat and other identifiers are different from the website URLs in the edition config

There seems to be some difference between the schema.org goals and RDF, so maybe schema.org data can be less precise.
For example, I was planning to add `owl:sameAs` to the RDF, and the URL `https://www.worldcat.org/oclc/@@@?tab=details`  IS NOT a WorldCat URI that would work there.   (especially with the query parameters)
